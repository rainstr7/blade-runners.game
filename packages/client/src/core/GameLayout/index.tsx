import React, { FC, useEffect, useRef } from 'react'
import styles from './style.module.css'
import useEvent from '../../hooks/useEvent'
import { Engine } from '../engine'

// TODO Необходимо сделать адаптивно
const engine = new Engine(1024, 768)

const GameLayout: FC = () => {
  useEvent('keydown', engine.handleKeyDown)
  useEvent('keyup', engine.handleKeyUp)

  const canvasRef = useRef<HTMLCanvasElement>(null)
  const lastTime = useRef(0)

  useEffect(() => {
    const canvas = canvasRef.current
    const context = canvas?.getContext('2d')

    if (!context) {
      throw new Error('Error getting context')
    }

    let animationFrameId = 0

    const render = (timeStamp: number) => {
      const deltaTime = timeStamp - lastTime.current
      lastTime.current = timeStamp

      engine.game(context, deltaTime)
      if (engine.gameOver) {
        return
      }
      animationFrameId = requestAnimationFrame(render)
    }
    render(0)

    return () => {
      cancelAnimationFrame(animationFrameId)
    }
  }, [engine.game])

  return (
    <div className={styles.container}>
      <canvas ref={canvasRef} width={1024} height={768} />
    </div>
  )
}

export default GameLayout
