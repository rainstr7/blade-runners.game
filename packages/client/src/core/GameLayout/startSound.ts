import audioSrc from '../../assets/sounds/start.mp3'

export default function play() {
  const audio = new Audio(audioSrc)
  audio.play().catch()
}
