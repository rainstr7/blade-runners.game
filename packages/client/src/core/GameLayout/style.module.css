.Container {
  position: relative;
  background-color: #440c77;
  width: 1024px;
  height: 768px;
}

.Canvas {
  -webkit-box-shadow: 0px 0px 20px 23px rgba(0,0,0,1);
  -moz-box-shadow: 0px 0px 20px 23px rgba(0,0,0,1);
  box-shadow: 0px 0px 20px 23px rgba(0,0,0,1);
  /*-webkit-box-shadow: inset 0px 0px 12px 6px rgba(0,0,0,1);*/
  /*-moz-box-shadow: inset 0px 0px 12px 6px rgba(0,0,0,1);*/
  /*box-shadow: inset 0px 0px 12px 6px rgba(0,0,0,1);*/
}
