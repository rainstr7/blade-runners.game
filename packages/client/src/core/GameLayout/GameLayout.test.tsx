import { render } from '@testing-library/react'
import GameLayout from '.'

test('корректный рендеринг GameLayout', () => {
  const { container } = render(<GameLayout />)

  expect(container).toMatchSnapshot()
})
