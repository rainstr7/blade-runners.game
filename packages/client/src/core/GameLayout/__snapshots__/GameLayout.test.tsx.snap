// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`корректный рендеринг GameLayout 1`] = `
<div>
  <div>
    <canvas
      height="768"
      width="1024"
    />
  </div>
</div>
`;
